#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Workaround invalid resolving of lazy evaluation stubs
# DP: Related bugs: #265678
# DP: Dpatch author: Thiemo Seufer <ths@debian.org>
# DP: Patch author: Thiemo Seufer <ths@debian.org>
# DP: Upstream status: Debian-Specific
# DP: Status Details: The correct fix will be done to modify toolchains in etch.
# DP: Date: 2005-04-11, 2005-04-16 updated by gotom

PATCHLEVEL=2

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
2005-04-16  GOTO Masanori  <gotom@debian.org>

	* sysdeps/mips/dl-machine.h: Drop due to _dl_lookup_versioned_symbol
	and _dl_lookup_symbol are changed into _dl_lookup_symbol_x.

2005-04-11  Thiemo Seufer <ths@debian.org>

	* first version.

diff -upr build-tree.orig/glibc-2.3.2/elf/do-lookup.h build-tree/glibc-2.3.2/elf/do-lookup.h
--- build-tree.orig/glibc-2.3.2/elf/do-lookup.h	2005-02-28 23:42:31.000000000 +0100
+++ build-tree/glibc-2.3.2/elf/do-lookup.h	2005-04-11 18:19:20.000000000 +0200
@@ -209,6 +209,13 @@ FCT (const char *undef_name, unsigned lo
 		}
 	      /* FALLTHROUGH */
 	    case STB_GLOBAL:
+#ifdef __mips__
+	      /* HACK: MIPS marks its lazy evaluation stubs with SHN_UNDEF
+		 symbols, we skip them. */
+	      if (sym->st_shndx == SHN_UNDEF)
+		break;
+#endif
+
 	      /* Global definition.  Just what we need.  */
 	      result->s = sym;
 	      result->m = map;
